File: BC-W4/Army/Spells/DemonsCall/DemonsCall.cpp
#include "DemonsCall.h"

DemonsCall::DemonsCall(int actionPoints, int manaCost, const std::string& spellName) : Spell(actionPoints, manaCost, spellName) {};
DemonsCall::~DemonsCall() {};

Demon* DemonsCall::action(Location* nLocation) {
    std::string name = "Slave" + demonAmount;
    
    Demon* demon = new Demon(name, nLocation->findFreeLocation().getX(), nLocation->findFreeLocation().getY());
    
    return demon;
    
    demonAmount += 1;
};

void DemonsCall::action(Unit* enemy) {
    throw InvalidActionException();
};

void DemonsCall::heal(Unit* unit) {
    throw InvalidHealException();
};

int DemonsCall::demonAmount = 1;

File: BC-W4/ArmyJava/src/spells/DemonsCall.java
package spells;

import exceptions.InvalidCastException;
import units.Demon;
import units.Unit;

public class DemonsCall extends Spell {
    private static int counter = 1;

    public DemonsCall() {
        super("DemonsCall", 50, 0, true);
    }

    public DemonsCall(int actionPoints) {
        super("DemonsCall", 50, actionPoints, true);
    }

    public Demon action() {
        String name = "Slave" + counter;
        Demon slave = new Demon(name);

        counter += 1;
        return slave;
    }
}
